src/backend/cdb/motion/cdbmotion.c

```diff
@@ -1274,3 +1277,77 @@ UpdateSentRecordCache(int32 *sent_record_typmod)
 		*sent_record_typmod = NextRecordTypmod;
 	}
 }
+
+void
+SetCurrentMotionIPCLayer(int new_type)
+{
+	int type;
+
+	/* interconnect.so is not loaded. */
+	if (CurrentIPCLayerImplNum == 0)
+	{
+		Assert(CurrentMotionIPCLayer == NULL);
+		Gp_interconnect_type = new_type;
+		ereport(WARNING,
+				(errcode(ERRCODE_WARNING_GP_INTERCONNECTION),
+				 errmsg("No any IPCLayer implement loaded.")));
+
+		return;
+	}
+
+	/* check new type */
+	if (IsICTypeExist(new_type))
+	{
+		type = new_type;
+	}
+	else
+	{
+		ereport(WARNING,
+				(errcode(ERRCODE_WARNING_GP_INTERCONNECTION),
+				 errmsg("No IPCLayer implement found with type: %d, choose default one.",
+						new_type)));
+
+		/* try best to find one appropriate implement */
+		if (IsICTypeExist(Gp_interconnect_type))
+		{
+			/* keep old setting */
+			type = Gp_interconnect_type;
+		}
+		else
+		{
+			if (IsICTypeExist(INTERCONNECT_TYPE_UDPIFC))
+			{
+				type = INTERCONNECT_TYPE_UDPIFC;
+			}
+			else
+			{
+				Assert(CurrentIPCLayerImplNum);
+				type = IPCLayerImpls[0]->ic_type;
+			}
+		}
```

Comment:
Please move the code to another function. Given the preferred order of ic types, choose the one in the order list.

Reply:
`Gp_interconnect_type` is dispatched from QD. Will it be possible that the chosen the interconnect types are inconsistent among nodes?

Reply:
fix: add TryFindICType() to find the ic type of the most appropriate implement;

Reply:
> Please move the code to another function. Given the preferred order of ic types, choose the one in the order list.

fix: add TryFindICType() to find the ic type of the most appropriate implement;

Reply:
> `Gp_interconnect_type` is dispatched from QD. Will it be possible that the chosen the interconnect types are inconsistent among nodes?

It's impossible.
```
BackendStartup
  BackendInitialize
    ProcessStartupPacket   save guc_opts from QD into port->guc_opts
BackendRun
  PostgresMain
    InitializeGUCOptions,  just run if IsUnderPostmaster == false
      InitializeOneGUCOption
    InitPostgres     
      process_startup_options  postinit.c:1193
         process_postgres_switches case 'c' or '-'
             SetConfigOption
                 set_config_option
      cdb_setup                        postinit.c:1266
          CurrentMotionIPCLayer->InitMotionLayerIPC()
```